# Spam Detector
### Author: Saikumar Kalyankrishnan

In [41]:
import pandas as pd

## Retrieve the Data

The data is located at [https://static.bc-edx.com/mbc/ai/m4/datasets/spam-data.csv](https://static.bc-edx.com/mbc/ai/m4/datasets/spam-data.csv)

Dataset Source: [UCI Machine Learning Library](https://archive-beta.ics.uci.edu/dataset/94/spambase)

Import the data using Pandas. Display the resulting DataFrame to confirm the import was successful.

In [42]:
# Import the data
data = pd.read_csv("http://static.bc-edx.com/mbc/ai/m4/datasets/spam-data.csv")
data.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  spam  
0                       278     1  
1                      1028     1  
2                      2259     1  
3                       191     1  
4                       191     1  

[5 rows x 58 columns]

## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct! 

Write down your prediction in the designated cells in your Jupyter Notebook, and provide justification for your educated guess.

<font color='yellow' size='4'>*_I believe Random Forest Model would perform better than Logistic Regression as it makes predictions based on random decision trees.
This prevents overfitting and provides a more accurate prediction especially with imbalanced data not being prone to biased decisions.
Random Forest Model can manage non-linear correlations between features and the label.
Run efficiently on large datasets._*</font>

## Split the Data into Training and Testing Sets

In [43]:
# Create the labels set `y` and features DataFrame `X`
# Import Module
from sklearn.model_selection import train_test_split

# Split training and testing sets
# Create the features DataFrame, X
X = data.copy()
X = X.drop(columns='spam')

# Create the target set `y`
y = data['spam']

In [44]:
# Check the balance of the labels variable (`y`) by using the `value_counts` function.
# Check the number of Spam vs. Non-Spam ('spam') using value_counts
# 1: Spam | 0: Non-Spam
y.value_counts()

spam
0    2788
1    1813
Name: count, dtype: int64

In [45]:
# Split the data into X_train, X_test, y_train, y_test using train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=99)

# Review the X_train DataFrame
X_train.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
1617            0.00               0.00           0.00           0.0   
419             0.51               0.43           0.29           0.0   
2378            0.00               0.00           1.46           0.0   
1925            0.00               0.00           0.00           0.0   
455             0.00               0.00           0.71           0.0   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
1617           0.23            0.00              0.00                0.00   
419            0.14            0.03              0.00                0.18   
2378           0.00            0.00              0.00                0.00   
1925           0.00            0.00              0.00                1.78   
455            0.17            0.00              0.35                0.35   

      word_freq_order  word_freq_mail  ...  word_freq_conference  char_freq_;  \
1617             0.00            0.00  ...                  0.11        0.000   
419              0.54            0.62  ...                  0.00        0.012   
2378             0.00            0.00  ...                  0.00        0.471   
1925             0.00            0.00  ...                  0.00        0.000   
455              0.00            0.17  ...                  0.00        0.000   

      char_freq_(  char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
1617        0.381          0.0        0.016        0.000        0.000   
419         0.078          0.0        0.443        0.510        0.133   
2378        0.550          0.0        0.078        0.000        0.000   
1925        0.000          0.0        0.000        0.000        0.000   
455         0.057          0.0        0.057        0.171        0.000   

      capital_run_length_average  capital_run_length_longest  \
1617                       2.470                          41   
419                        6.590                         739   
2378                       2.552                          16   
1925                       1.666                           7   
455                        1.974                          34   

      capital_run_length_total  
1617                       504  
419                       2333  
2378                       171  
1925                        15  
455                        229  

[5 rows x 57 columns]

In [46]:
# Review the y_train DataFrame
y_train.head()

1617    1
419     1
2378    0
1925    0
455     1
Name: spam, dtype: int64

## Scale the Features

Use the `StandardScaler` to scale the features' data. Remember that only `X_train` and `X_test` DataFrames should be scaled.

In [47]:
from sklearn.preprocessing import StandardScaler

# Create the StandardScaler instance
scaler = StandardScaler()
scaler

StandardScaler()

In [48]:
# Fit the Standard Scaler with the training data
X_scaler = scaler.fit(X_train)

In [49]:
# Scale the training data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Create and Fit a Logistic Regression Model

Create a Logistic Regression model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score.

In [50]:
# Train a Logistic Regression model and print the model score
from sklearn.linear_model import LogisticRegression

# Create a `LogisticRegression` function and assign it to a variable named `logistic_regression_model`.
logistic_regression_model = LogisticRegression(random_state=1)

# Fit the model
logistic_regression_model.fit(X_train_scaled, y_train)

# Score the model
print(f"Training Data Score: {logistic_regression_model.score(X_train_scaled, y_train)}")
print(f"Testing Data Score: {logistic_regression_model.score(X_test_scaled, y_test)}")

Training Data Score: 0.9278260869565217
Testing Data Score: 0.9244135534317984


In [51]:
# Make and save testing predictions with the saved logistic regression model using the test data
testing_predictions = logistic_regression_model.predict(X_test_scaled)

# Review the predictions
testing_predictions

array([0, 0, 0, ..., 0, 1, 0], dtype=int64)

In [52]:
# Save both the test predictions and actual test values to a DataFrame
results_data = pd.DataFrame({
    "Testing Data Predictions": testing_predictions, 
    "Testing Data Actual Targets": y_test})

results_data

Testing Data Predictions  Testing Data Actual Targets
893                          0                            1
2279                         0                            0
3691                         0                            0
4256                         0                            0
3968                         0                            0
...                        ...                          ...
1169                         1                            1
2769                         0                            0
1710                         0                            1
1459                         1                            1
4577                         0                            0

[1151 rows x 2 columns]

In [53]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
from sklearn.metrics import accuracy_score

# Calculate the model's accuracy on the test dataset
acc_score = accuracy_score(y_test, testing_predictions)

print(f"Accuracy for Logistic Regression Model with Scaled Data: {acc_score}")
acc_score

Accuracy for Logistic Regression Model with Scaled Data: 0.9244135534317984


0.9244135534317984

## Create and Fit a Logistic Regression Model with Non-Scaled Data

In [54]:
# Train a Logistic Regression model and print the model score
from sklearn.linear_model import LogisticRegression

# Create a `LogisticRegression` function and assign it to a variable named `logistic_regression_model`.
logistic_regression_non_scaled = LogisticRegression(max_iter=2000)

# Fit the model
logistic_regression_non_scaled.fit(X_train.values, y_train)

# Score the model
print(f"Training Data Score: {logistic_regression_non_scaled.score(X_train.values, y_train)}")
print(f"Testing Data Score: {logistic_regression_non_scaled.score(X_test.values, y_test)}")

# Make and save testing predictions with the saved logistic regression model using the test data
testing_predictions_non_scaled = logistic_regression_model.predict(X_test.values)

# Calculate the model's accuracy on the test dataset
acc_score = accuracy_score(y_test, testing_predictions_non_scaled)

print(f"Accuracy for Logistic Regression Model without Scaled Data: {acc_score}")
acc_score

Training Data Score: 0.9295652173913044
Testing Data Score: 0.9261511728931364
Accuracy for Logistic Regression Model without Scaled Data: 0.5030408340573415


0.5030408340573415

## Create and Fit a Random Forest Classifier Model

Create a Random Forest Classifier model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score.

In [55]:
# Train a Random Forest Classifier model and print the model score
from sklearn.ensemble import RandomForestClassifier
rf_model = RandomForestClassifier(n_estimators=128, random_state=1, criterion='entropy')

# Fit the model
rf_model = rf_model.fit(X_train, y_train)

In [56]:
# Make and save testing predictions with the saved Random Forest Classifier model using the test data
testing_predictions = rf_model.predict(X_test)

# Review the predictions
testing_predictions

array([0, 0, 0, ..., 1, 1, 0], dtype=int64)

In [57]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
acc_score = accuracy_score(y_test, testing_predictions)

# Display results
print(f"Accuracy for Random Forest Classifier Model without Scaled Data: {acc_score}")
acc_score

Accuracy for Random Forest Classifier Model without Scaled Data: 0.9574283231972198


0.9574283231972198

In [58]:
# Get the feature importance array
importances = rf_model.feature_importances_

# List the top 10 most important features
importances_sorted = sorted(zip(rf_model.feature_importances_, X.columns), reverse=True)
importances_sorted[:10]

[(0.10862296291345132, 'char_freq_!'),
 (0.08502471094582552, 'char_freq_$'),
 (0.07084893700258087, 'capital_run_length_average'),
 (0.07066424603146788, 'word_freq_remove'),
 (0.06334277292805543, 'capital_run_length_longest'),
 (0.05205276070762099, 'word_freq_hp'),
 (0.04901741863213349, 'word_freq_free'),
 (0.04684202268029729, 'capital_run_length_total'),
 (0.04179646948577686, 'word_freq_your'),
 (0.03247748216789178, 'word_freq_money')]

## Create and Fit a Random Forest Classifier Model with Scaled Data

In [59]:
# Train a Random Forest Classifier model and print the model score for scaled data
rf_scaled = RandomForestClassifier(n_estimators=128, random_state=1, criterion='entropy')

# Fit the model
rf_scaled = rf_scaled.fit(X_train_scaled, y_train)

# Make and save testing predictions with the saved Random Forest Classifier model using the test data
testing_predictions_scaled = rf_scaled.predict(X_test_scaled)

# Review the predictions for scaled data
testing_predictions_scaled

array([0, 0, 0, ..., 1, 1, 0], dtype=int64)

In [60]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions_scaled`.
acc_score = accuracy_score(y_test, testing_predictions_scaled)

# Display results
print(f"Accuracy for Random Forest Classifier Model with Scaled Data: {acc_score}")
acc_score

Accuracy for Random Forest Classifier Model with Scaled Data: 0.9582971329278888


0.9582971329278888

## Evaluate the Models

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the following markdown cell.

<font color='yellow' size='4'>As predicted, Random Forest model did perform better than Logistic Regression, but a few observations spiked my curiosity.
1.The factor of `random_state`: the value did have an effect on the accuracy for both the models
2.The `n_estimators` also had an effect on the accuracy of Random Forest model along with the parameter `criterion` (`criterion=entropy` had the maximum accuracy)
3.Is scaling recommended in this case where the accuracy isn't improving too much?<br>
The challenge would be to find the perfect combination of `random_state`, `n_estimators` and `criterion` to improve the accuracy of the model.</font>
<br>

---

### Logistic Regression Model: Observations

| random_state | scaled |                   accuracy                    |
|:------------:|:------:|:---------------------------------------------:|
|      78      |   Y    | <font color='green'>0.9278887923544744</font> |
|      99      |   Y    |              0.9244135534317984               |
|     151      |   Y    |              0.9226759339704604               |
|      56      |   Y    |              0.9244135534317984               |
|   default    |   Y    |              0.9261511728931364               |
|   default    |   N    |              0.5030408340573415               |

<br>
<br>

---

### Random Forest Classifier Model: Observations
| random_state | scaled |                   accuracy                    |
|:------------:|:------:|:---------------------------------------------:|
|      1       |   N    |              0.9574283231972198               |
|      1       |   Y    | <font color='green'>0.9582971329278888</font> |
<br>

| n_estimators |                   accuracy                    |
|:------------:|:---------------------------------------------:|
|     128      | <font color='green'>0.9600347523892268</font> |
|      64      |              0.9591659426585578               |
|   default    |              0.9565595134665508               |
<br>

| criterion |                   accuracy                    |
|:---------:|:---------------------------------------------:|
|  entropy  | <font color='green'>0.9600347523892268</font> |
|   gini    |              0.9556907037358818               |
| log_loss  |              0.9548218940052129               |
